ENH: Fix compile warnings in FilterUtilities. Init array in DataStore when resizing

imikejackson opened the pull request:
FilterUtilities had a function without a return statement

- DataStore did not attempt to initialize the array with any value. This leads to difficulty debugging when the values are random or happening randomly (if the re-allocation happens to be in the same place as a once-valid array.

This change attempts to allow for the initialization value to be used. In case *no* initialization value is passed into the constructor then a default 0xABAB byte pattern is used instead.

imikejackson committed:
ENH: Fix compile warnings in FilterUtilities. Init array in DataStore when resizing

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

JDuffeyBQ commented on src/simplnx/Utilities/FilterUtilities.hpp:
```diff
@@ -94,6 +95,9 @@ auto RunTemplateClass(DataType dataType, ArgsT&&... args)
       return ClassT<float64>(std::forward<ArgsT>(args)...)();
     }
   }
+  std::stringstream ss;
+  ss << "FilterUtilities::RunTemplateClass<> Error: dataType did not match any known type. DataType was " << to_underlying(dataType);
+  throw std::runtime_error(ss.str());
```
```suggestion
  throw std::runtime_error(fmt::format("FilterUtilities::RunTemplateClass<> Error: dataType did not match any known type. DataType was {}", DataTypeToString(dataType)));
```

JDuffeyBQ commented on src/simplnx/DataStructure/DataStore.hpp:
```diff
@@ -82,6 +83,47 @@ class DataStore : public AbstractDataStore<T>
   , m_NumTuples(std::accumulate(m_TupleShape.cbegin(), m_TupleShape.cend(), static_cast<size_t>(1), std::multiplies<>()))
   {
+    // Because no init value is passed into the constructor, we will use a "mudflap" style value that is easy to debug.
+    if constexpr(sizeof(T) == 1)
```
```cpp
template <class T>
constexpr T GetMudflap() noexcept
{
  if constexpr(sizeof(T) == 1)
  {
    return std::bit_cast<T>(static_cast<uint8>(0xAB));
  }
  if constexpr(sizeof(T) == 2)
  {
    return std::bit_cast<T>(static_cast<uint16>(0xABAB));
  }
  if constexpr(sizeof(T) == 4)
  {
    return std::bit_cast<T>(static_cast<uint32>(0xABABABAB));
  }
  if constexpr(sizeof(T) == 8)
  {
    return std::bit_cast<T>(static_cast<uint64>(0xABABABABABABABAB));
  }
}
```
Something like this would work.

imikejackson committed:
Optimize code based on feedback.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Update src/simplnx/Utilities/FilterUtilities.hpp

Co-authored-by: Jared Duffey <jared.duffey@bluequartz.net>

imikejackson commented:
Requires https://github.com/BlueQuartzSoftware/simplnx/pull/1028

nyoungbq commented on src/simplnx/Common/TypesUtility.hpp:
```diff
@@ -10,6 +10,33 @@
 
```
```suggestion
#include <bit>

```

nyoungbq commented on src/simplnx/Utilities/FilterUtilities.hpp:
```diff
@@ -2,6 +2,7 @@
 
 #include "ParallelAlgorithmUtilities.hpp"
 #include "simplnx/Common/Result.hpp"
+#include "simplnx/Common/TypeTraits.hpp"
```
```suggestion
#include "simplnx/Common/TypeTraits.hpp"
#include "simplnx/Common/TypeUtilities.hpp"
```

imikejackson committed:
Fix missing include

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Update src/simplnx/Utilities/FilterUtilities.hpp

Co-authored-by: Nathan Young <nathan.young@bluequartz.net>

imikejackson committed:
Update src/simplnx/Common/TypesUtility.hpp

Co-authored-by: Nathan Young <nathan.young@bluequartz.net>

JDuffeyBQ committed:
Fixed clang format

Signed-off-by: Jared Duffey <jared.duffey@bluequartz.net>

JDuffeyBQ committed:
Fixed include

Signed-off-by: Jared Duffey <jared.duffey@bluequartz.net>

imikejackson committed:
Changes to compile under Xcode 14.2.

#include <bit> with bit_cast<> was not instroduced until Xcode 14.3 which requires
macOS 13.x (Ventura)

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>
